test/controllers/tasks_controller_test.rb

```diff
@@ -71,56 +69,133 @@
       expect(new_task.description).must_equal task_hash[:task][:description]
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
-
+      expect(new_task.completion_date.to_i).must_equal task_hash[:task][:completion_date].to_i
+      
+      
       must_respond_with :redirect
-      must_redirect_to task_path(new_task.id)
+      must_redirect_to tasks_path
     end
   end
-
+  
   # Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      
+      # Act
+      get edit_task_path(task.id)
+      
+      # Assert
+      must_respond_with :success
     end
-
+    
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      
+      get edit_task_path(-12)
+      
+      # Assert
+      must_respond_with :redirect
+      expect(flash[:error]).must_equal "Could not find task with id: -12"
     end
   end
-
+  
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      # Your code here
+      
+      old_task = Task.find_by(id: task.id)
+      
+      task_hash = {
+        task: {
+          name: "sample task",
+          description: "this is a test",
+          completion_date: nil,
+        },
+      }
+      
+      expect {patch task_path(old_task.id), 
+        params: task_hash}.wont_change "Task.count"
+        
+        
+        
+      end
+      
+      it "will redirect to the root page if given an invalid id" do
+        patch task_path(-12)
+        
+        # Assert
+        must_respond_with :redirect
+        expect(flash[:error]).must_equal "Could not find task with id: -12"
+      end
     end
-
-    it "will redirect to the root page if given an invalid id" do
-      # Your code here
+    
+    # Complete these tests for Wave 4
+    describe "destroy" do
+      
+      it "can destroy an existing task" do
+        
+        old_task = Task.find_by(id: task.id)
+        
+        
+        expect {delete task_path(old_task.id)}.must_change "Task.count", -1
+        
+        
+        
+      end
+      
+      it "will redirect to the root page if given an invalid id" do
+        delete task_path(-12)
+        
+        # Assert
+        must_respond_with :redirect
+        expect(flash[:error]).must_equal "Could not find task with id: -12"
+      end
+      
     end
-  end
-
-  # Complete these tests for Wave 4
-  describe "destroy" do
-    # Your tests go here
-
-  end
-
-  # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
-  end
-end
+    
+    # Complete for Wave 4
+    describe "toggle_complete" do
```

Comment:
Fantastic tests for the `toggle_complete` functionality! The last remaining edge case is trying to do this with an invalid id.